### PR TITLE
code: suggestions for #2169

### PR DESCRIPTION
**Pull request by noomorph**
@DmytroKost ,  I think there are too many redundant classes at the end of the day.

1. I suggest moving shared functionality inside the base AndroidDriver, e.g. allocating the devices.
2. Implement different device allocation using the class hierarchy you created, but without that small (for-of) loop wrapper utility class .

I have issues on my laptop at the moment with getting Android emulators work at the moment, and not all unit tests work, but in general this is a direction I suggest you to take. We'll spare about -100 redundant lines by reusing functionality, not copying it, I think.

### CODE REVIEW

File: detox/src/devices/drivers/android/tools/FreeDeviceFinderBase.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    return null;
+  }
+
+  /** @protected */
+  async isDeviceMatching(candidate, matcher) {
+    return false;
+  }
+}
+
+module.exports = FreeDeviceFinderBase;
```
Comment (d4vidi) [inline detox/src/devices/drivers/android/tools/FreeDeviceFinderBase.js:37]: I highly approve of this 👍

File: detox/src/devices/drivers/android/AndroidDriver.js [hunk 2]
```diff
@@ -58,6 +64,16 @@ class AndroidDriver extends DeviceDriverBase {
     };
   }
 
+  /** @protected */
+  async allocateDevice(matcher) {
+    log.debug({ event: 'ALLOCATE_DEVICE' }, `Trying to allocate a device based on %j`, matcher);
+    const adbName = await this.deviceRegistry.allocateDevice(() => {
+      return this.freeDeviceFinder.findFreeDevice(matcher);
```
Comment (d4vidi) [inline detox/src/devices/drivers/android/AndroidDriver.js]: Not very fond of this part -- I don't appreciate the base implementation imposing this logic. Better keep the `freeDeviceFinder` at the implementor's side, even with the cost of (very little) code duplication, if you ask me.
I'm nevertheless okay with the skeletal implementation here, but perhaps with the subtle replacement of `return this.freeDeviceFinder.findFreeDevice(matcher)` with `return this._someAbstractMethod()`

File: detox/src/devices/drivers/android/AttachedAndroidDriver.js [hunk 3]
```diff
@@ -39,19 +34,6 @@ class AttachedAndroidDriver extends AndroidDriver {
     await this.deviceRegistry.disposeDevice(adbName);
     await super.cleanup(adbName, bundleId);
   }
-
-  async _allocateDevice(adbNamePattern) {
-    log.debug({ event: ALLOCATE_DEVICE_LOG_EVT }, `Trying to allocate a device based on "${adbNamePattern}"`);
-    const adbName = await this.deviceRegistry.allocateDevice(() => this._doAllocateDevice(adbNamePattern));
-    log.debug({ event: ALLOCATE_DEVICE_LOG_EVT }, `Settled on ${adbName}`);
-    return adbName;
-  }
```
Comment (d4vidi) [inline detox/src/devices/drivers/android/AttachedAndroidDriver.js:34]: yes, much better

File: detox/src/devices/drivers/android/AndroidDriver.js (hunk 2 above)
Comment (noomorph) [reply detox/src/devices/drivers/android/AndroidDriver.js]: I see. @d4vidi, do you mean that `this.freeDeviceFinder = null;` in the base class looks somewhat ugly?
Well, I tend to agree — both implementors of AndroidDriver can handle its initialization completely, without the base `AndroidDriver` class knowing those details, if we make `findFreeDevice` method abstract. 👍

Comment (d4vidi) [reply detox/src/devices/drivers/android/AndroidDriver.js]: yes, exactly

**Comment by noomorph** [comment]
@d4vidi, thanks. Fixed in https://github.com/DmytroKost/Detox/pull/1/commits/f950fa73e3ea1546554fdfbb626d29ca3b12cbaf 

Still there are 2 unit tests suites failing, hope they are very trivial to solve though, @DmytroKost .

**Comment by DmytroKost** [comment]
@noomorph Oh... I've been working on the suggested changes as well and just finished refactoring. I think I'm going to push the changes to my branch with the fixed unit tests. The changes are not exactly the same, though. I don't know which version you prefer.

**Comment by noomorph** [comment]
@DmytroKost, oh, I thought you indeed have frozen your works till the consensus. Anyway, let me know where to look, and we'll decide right there.

**Comment by DmytroKost** [comment]
@noomorph That's okay. My changes are very close to your suggestions. Firstly, I'm going to verify that parallel testing works on Genymotion after my refactoring. Then I'll push the changes with a detailed description.

**Comment by DmytroKost** [comment]
@noomorph @d4vidi I pushed new changes to my [branch](https://github.com/DmytroKost/Detox/tree/uic/adb-name-array)

### File structure

-  src/devices/drivers
    -  DeviceDriverBase.js
    -  android
        -  AndroidDriver.js
        -  attached
            -  AttachedAndroidDriver.js
            -  FreeAdbDeviceFinder.js
        -  emulator
            -  EmulatorDriver.js
            -  FreeEmulatorFinder.js

### Call hierarchy

DriverRegistry resolves:

-  'android.emulator': EmulatorDriver
    -  EmulatorDriver.acquireFreeDevice
    -  AndroidDriver.allocateDevice
    -  ABSTRACT AndroidDriver.doAllocateDevice -> IMPL EmulatorDriver.doAllocateDevice
    -  FreeEmulatorFinder.findFreeDevice -> SUPER FreeDeviceFinderBase.findFreeDevice
    -  ABSTRACT FreeDeviceFinderBase.isDeviceMatching -> IMPL FreeEmulatorFinder.isDeviceMatching
-  'android.attached': AttachedAndroidDriver
    -  AttachedAndroidDriver.acquireFreeDevice
    -  AndroidDriver.allocateDevice
    -  ABSTRACT AndroidDriver.doAllocateDevice -> IMPL AttachedAndroidDriver.doAllocateDevice
    -  FreeAdbDeviceFinder.findFreeDevice -> SUPER FreeDeviceFinderBase.findFreeDevice
    -  ABSTRACT FreeDeviceFinderBase.isDeviceMatching -> IMPL FreeAdbDeviceFinder.isDeviceMatching

#### Other changes

Difference from the suggestions in https://github.com/DmytroKost/Detox/pull/1 should be minimal:

-  `AndroidDriver.cleanup()` now calls `this.deviceRegistry.disposeDevice`, so no need to override it in subclasses;
-  `EmulatorDriver` will boot a new emulator with `_createDevice` if no free emulators found;

#### Other notes

-  Rebased on the latest master
-  All unit tests are green with 100% coverage (locally)
-  Running tests on multiple Genymotion emulators still  works after the refactoring


Unresolved discussion about examples https://github.com/wix/Detox/pull/2169#discussion_r458667027

**Pull request closed without merging**